### ISSUE TITLE
Activate features of individual crates in a cargo workspace

### ISSUE THREAD

**Issue by Lucretiel** [rust-lang/rust-analyzer#10298]
Currently, the only two options for cargo features in rust-analyzer are "all features" and "list of features". If you're in a cargo workspace, the "list of features" seems to apply to all crates, which breaks certain workflows when working in those crates (in particular, the "▷Run Test" inline buttons). There should be a way to apply designate specific crates to apply features in the project settings.

**Comment by ia0** [comment]
Note that there's the same issue with `cargo.noDefaultFeatures` which doesn't work in a cargo workspace.

**Comment by shadeglare** [comment]
As I understand the CLion Rust plugin has a similar ability. You can activate certain features in a crate during development.
Even if you make a reference to the crate in another crate the with different feature flags you edit the crate with the IDE pinned features of your choice.
<img width="258" alt="Screenshot 2022-07-31 at 13 14 52" src="https://user-images.githubusercontent.com/995484/182021676-91243d30-6fb4-402a-802a-430af99a965d.png">

**Comment by dhedey** [comment]
Related to https://github.com/rust-lang/rust-analyzer/issues/8521

This is a big problem for us - we have a few mutually incompatible features (for relatively legitimate reasons) -- and a workspace with many crates with different features. But enabling any feature with eg:

```
    "rust-analyzer.cargo.features": [
        "serde",
    ],
```

Works fine when running analysis over the whole workspace*, but errors when running tests under a particular project which doesn't support that flag. As a workaround, you can add an empty dummy feature to each individual cargo project, but it's not ideal.

* this is not strictly true - using `rust-analyzer.linkedProjects`, each individual project needs to have that feature.

The following would all be sufficient fixes for us:
* Use the requested features only for projects / workspaces / crates where they are available.
* Allow us to choose features by Cargo.toml file - ie allow a mapping of Cargo.toml paths to features.
* Allow us to choose features through the UI like in CLion

**Comment by Veykril** [comment]
https://github.com/rust-lang/rust-analyzer/pull/14910 addresses this for single workspace projects. Multi workspaces still suffer here

**Comment by Lucretiel** [comment]
> #14910 addresses this for single workspace projects. Multi workspaces still suffer here

Does it? That just prevents feature `abc` from being enabled in crates that don't have it; it doesn't provide any way to enable `abc` in some crates but not others.

**Comment by Veykril** [comment]
> which breaks certain workflows when working in those crates (in particular, the "▷Run Test" inline buttons)

It fixes this scenario, I didn't see an ask for enabling per crate mightve missed that.

**Comment by ChayimFriedman2** [comment]
You can enable per-package feature with `package/feature`. This is a Cargo syntax and rust-analyzer does not even need to know about it.

**Comment by nasso** [comment]
correct me if im wrong, but i believe rust-analyzer [sometimes calls the check command with `-p <crate>`](https://github.com/rust-lang/rust-analyzer/blob/64ddb549bc9a70d011328746fa46a8883f937b6b/crates/rust-analyzer/src/flycheck.rs#L898), which breaks when using `rust-analyzer.{cargo,check}.features` to select features

e.g. this is an error im getting:
```
ERROR Flycheck failed to run the following command: CommandHandle { program: "/nix/store/64c81dfywyyvicwc5djrb3s2zl7qggma-rust-default-1.89.0/bin/cargo", arguments: ["clippy", "-p", "<redacted-crate>", "--bin", "<redacted-bin>", "--message-format=json", "--manifest-path", "<redacted-path>/Cargo.toml", "--keep-going", "--all-targets", "--features", "<redacted-crate>/<redacted-feature> <redacted-crate>/<redacted-feature>"], current_dir: Some("<redacted-path>") }, error=Cargo watcher failed, the command produced no valid metadata (exit code: ExitStatus(unix_wait_status(25856))):
error: the package '<redacted-crate>' does not contain these features: <redacted-crate>/<redacted-feature> <redacted-crate>/<redacted-feature>
```

**Comment by ChayimFriedman2** [comment]
Right, so we do need r-a to support the `package/feature` syntax and not pass the feature for other packages.

**Comment by ChayimFriedman2** [comment]
Implemented in https://github.com/rust-lang/rust-analyzer/pull/22432.
